apps/swiss/whoami.c

```diff
@@ -0,0 +1,206 @@
+//
+
+#include <minoca/lib/types.h>
+
+#include <assert.h>
+#include <getopt.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "swlib.h"
+
+//
+// ---------------------------------------------------------------- Definitions
+//
+
+#define WHOAMI_VERSION_MAJOR 1
+#define WHOAMI_VERSION_MINOR 0
+
+#define WHOAMI_USAGE                                                        \
+    "usage: whoami [options]\n"                                              \
+    "Print the user name associated with the current effective user ID. \n" \
+    "Same as id -un.\n"                                                     \
+    "Options are:\n"                                                        \
+    "  --help -- Show this help text and exit.\n"                           \
+    "  --version -- Print the application version information and exit.\n"
+
+#define WHOAMI_OPTIONS_STRING "h"
+
+//
+// ------------------------------------------------------ Data Type Definitions
+//
+
+//
+// ----------------------------------------------- Internal Function Prototypes
+//
+
+VOID
+WhoamiPrintUserName (
+    uid_t UserId
+    );
+
+//
+// -------------------------------------------------------------------- Globals
+//
+
+struct option WhoamiLongOptions[] = {
+    {"help", no_argument, 0, 'h'},
+    {"version", no_argument, 0, 'V'},
+    {NULL, 0, 0, 0},
+};
+
+//
+// ------------------------------------------------------------------ Functions
+//
+
+INT
+WhoamiMain (
+    INT ArgumentCount,
+    CHAR **Arguments
+    )
+
+/*++
+
+Routine Description:
+
+    This routine is the main entry point for the whoami utility.
+
+Arguments:
+
+    ArgumentCount - Supplies the number of command line arguments the program
+        was invoked with.
+
+    Arguments - Supplies a tokenized array of command line arguments.
+
+Return Value:
+
+    Returns an integer exit code. 0 for success, nonzero otherwise.
+
+--*/
+
+{
+
+    INT Option;
+    uid_t EffectiveUserId;
+    int Status;
+
+    //
+    // Process the control arguments.
+    //
+
+    while (TRUE) {
+        Option = getopt_long(ArgumentCount,
+                             Arguments,
+                             WHOAMI_OPTIONS_STRING,
+                             WhoamiLongOptions,
+                             NULL);
+
+        if (Option == -1) {
+            break;
+        }
+
+        if ((Option == '?') || (Option == ':')) {
+            Status = 1;
+            goto MainEnd;
+        }
+
+        switch (Option) {
+        case 'V':
+            SwPrintVersion(WHOAMI_VERSION_MAJOR, WHOAMI_VERSION_MINOR);
+            return 1;
+
+        case 'h':
+            printf(WHOAMI_USAGE);
+            return 1;
+
+        default:
+
+            assert(FALSE);
+
+            Status = 1;
+            goto MainEnd;
+        }
+    }
+
+    Status = 0;
+    EffectiveUserId = SwGetEffectiveUserId();
+    WhoamiPrintUserName(EffectiveUserId);
+
```

Comment:
I always try to follow the style and was about to ask you guys where I could find some spec. Thanks for the pointer.